filebeat/reader/line/line.go

```diff
@@ -18,177 +18,258 @@
 func (r *Reader) Next() ([]byte, int, error) {
-	// This loop is need in case advance detects an line ending which turns out
-	// not to be one when decoded. If that is the case, reading continues.
-	for {
-		// read next 'potential' line from input buffer/reader
-		err := r.advance()
-		if err != nil {
-			return nil, 0, err
-		}
+	return r.lineScanner.scan()
+}
 
-		// Check last decoded byte really being '\n' also unencoded
-		// if not, continue reading
-		buf := r.outBuffer.Bytes()
+type decoderReader struct {
+	in         io.Reader
+	decoder    transform.Transformer
+	buf        *streambuf.Buffer
+	encodedBuf *streambuf.Buffer
+	bufferSize int
+	symlen     []int
 
-		// This can happen if something goes wrong during decoding
-		if len(buf) == 0 {
-			logp.Err("Empty buffer returned by advance")
-			continue
-		}
+	offset      int
+	bytesOffset int
+}
 
-		if buf[len(buf)-1] == '\n' {
-			break
-		} else {
-			logp.Debug("line", "Line ending char found which wasn't one: %s", buf[len(buf)-1])
-		}
+// GetState returns the state of this and the previous readers
+func (r *Reader) GetState() common.MapStr {
+	return common.MapStr{
+		"decoder": common.MapStr{
+			"offset": r.lineScanner.in.offset,
+			"bytes":  r.lineScanner.in.bytesOffset,
+		},
+		"scanner": common.MapStr{
+			"offset":       r.lineScanner.offset,
+			"bytes":        r.lineScanner.bytesOffset,
+			"last_message": r.lineScanner.lastMessageLen,
+		},
 	}
+}
 
-	// output buffer contains complete line ending with '\n'. Extract
-	// byte slice from buffer and reset output buffer.
-	bytes, err := r.outBuffer.Collect(r.outBuffer.Len())
-	r.outBuffer.Reset()
-	if err != nil {
-		// This should never happen as otherwise we have a broken state
-		panic(err)
+func newDecoderReader(in io.Reader, codec encoding.Encoding, bufferSize int) *decoderReader {
+	return &decoderReader{
+		in:          in,
+		decoder:     codec.NewDecoder(),
+		buf:         streambuf.New(nil),
+		encodedBuf:  streambuf.New(nil),
+		bufferSize:  bufferSize,
+		offset:      0,
+		bytesOffset: 0,
 	}
-
-	// return and reset consumed bytes count
-	sz := r.byteCount
-	r.byteCount = 0
-	return bytes, sz, nil
 }
 
-// Reads from the buffer until a new line character is detected
-// Returns an error otherwise
-func (r *Reader) advance() error {
-	// Initial check if buffer has already a newLine character
-	idx := r.inBuffer.IndexFrom(r.inOffset, r.nl)
+func (r *decoderReader) read(buf []byte) (int, error) {
+	b := make([]byte, r.bufferSize)
 
-	// fill inBuffer until '\n' sequence has been found in input buffer
-	for idx == -1 {
-		// increase search offset to reduce iterations on buffer when looping
-		newOffset := r.inBuffer.Len() - len(r.nl)
-		if newOffset > r.inOffset {
-			r.inOffset = newOffset
-		}
+	if r.buf.Len() != 0 {
+		return r.copyToOut(buf)
+	}
 
-		buf := make([]byte, r.bufferSize)
+	for {
+		start := r.encodedBuf.Len()
+		n, err := r.in.Read(b[start:])
+		if err != nil {
+			return len(buf[:n]), err
```

Comment:
This return does not look correct. Its equivalent to `return n, err`. We will need a global variable tracking the actual number of bytes copied into buf.